### ISSUE TITLE
grid from many sources blows MoBIE up

### ISSUE THREAD

**Issue by martinschorb** [mobie/mobie-viewer-fiji#478]
Hi,

I started implementing the gallery views for our centrioles:

the view `crop_test1` in branch `grid-test2`contains:
- 235 crops 
- from 63 sources (several GB each)
- they should be displayed next to each other

I get:
```
https://raw.githubusercontent.com/mobie/centrioles-tomo-datasets/grid-test2/data/tomo/tables/crop_test1/default.tsv
Exception in thread "Fetcher-2" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-2" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-2" java.lang.OutOfMemoryError: Java heap space
	at com.sun.crypto.provider.CipherCore.prepareInputBuffer(CipherCore.java:1025)
	at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:919)
	at com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:491)
	at javax.crypto.Cipher.doFinal(Cipher.java:2376)
	at sun.security.ssl.CipherBox.decrypt(CipherBox.java:461)
	at sun.security.ssl.InputRecord.decrypt(InputRecord.java:172)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1017)
	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:933)
	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at org.apache.http.impl.io.SessionInputBufferImpl.read(SessionInputBufferImpl.java:175)
	at org.apache.http.impl.io.ContentLengthInputStream.read(ContentLengthInputStream.java:137)
	at org.apache.http.conn.EofSensorInputStream.read(EofSensorInputStream.java:118)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.event.ProgressInputStream.read(ProgressInputStream.java:158)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.event.ProgressInputStream.read(ProgressInputStream.java:158)
	at java.security.DigestInputStream.read(DigestInputStream.java:124)
	at com.amazonaws.services.s3.internal.DigestValidationInputStream.read(DigestValidationInputStream.java:47)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.services.s3.internal.S3AbortableInputStream.read(S3AbortableInputStream.java:99)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at org.janelia.saalfeldlab.n5.s3.N5AmazonS3Reader$S3ObjectInputStreamDrain.read(N5AmazonS3Reader.java:89)
	at java.io.InputStream.read(InputStream.java:179)
	at org.apache.commons.compress.utils.CountingInputStream.read(CountingInputStream.java:56)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
Exception in thread "Fetcher-3" Exception in thread "Fetcher-3" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-3" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-1" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-0" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-2" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-0" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-2" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-1" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-0" java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-2" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-1" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-0" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-2" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-1" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-0" Exception in thread "Fetcher-3" Exception in thread "Fetcher-0" Exception in thread "Fetcher-1" java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
	at com.sun.crypto.provider.CipherCore.prepareInputBuffer(CipherCore.java:1025)
	at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:919)
	at com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:491)
	at javax.crypto.Cipher.doFinal(Cipher.java:2376)
	at sun.security.ssl.CipherBox.decrypt(CipherBox.java:461)
	at sun.security.ssl.InputRecord.decrypt(InputRecord.java:172)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1017)
	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:933)
	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at org.apache.http.impl.io.SessionInputBufferImpl.read(SessionInputBufferImpl.java:175)
	at org.apache.http.impl.io.ContentLengthInputStream.read(ContentLengthInputStream.java:137)
	at org.apache.http.conn.EofSensorInputStream.read(EofSensorInputStream.java:118)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.event.ProgressInputStream.read(ProgressInputStream.java:158)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.event.ProgressInputStream.read(ProgressInputStream.java:158)
	at java.security.DigestInputStream.read(DigestInputStream.java:124)
	at com.amazonaws.services.s3.internal.DigestValidationInputStream.read(DigestValidationInputStream.java:47)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at com.amazonaws.services.s3.internal.S3AbortableInputStream.read(S3AbortableInputStream.java:99)
	at com.amazonaws.internal.SdkFilterInputStream.read(SdkFilterInputStream.java:84)
	at org.janelia.saalfeldlab.n5.s3.N5AmazonS3Reader$S3ObjectInputStreamDrain.read(N5AmazonS3Reader.java:89)
	at java.io.InputStream.read(InputStream.java:179)
	at org.apache.commons.compress.utils.CountingInputStream.read(CountingInputStream.java:56)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
Exception in thread "Fetcher-0" java.lang.OutOfMemoryError: Java heap space
Error loading setup0/timepoint0/s0 at block [5, 4, 2]: java.lang.IllegalArgumentException: Self-suppression not permitted
Exception in thread "Fetcher-0" java.lang.OutOfMemoryError: Java heap space
Exception in thread "Fetcher-0" java.lang.OutOfMemoryError: Java heap space
Error loading setup0/timepoint0/s0 at block [4, 4, 7]: java.lang.IllegalArgumentException: Self-suppression not permitted
Exception in thread "Fetcher-1" java.lang.OutOfMemoryError: Java heap space
```
Is the downsampling not enough to load all the needed chunks?
How can we get around this?

If I load the untransformed sources into a full grid (`B-ALL_tomos`), I do not even see the dat, only very tiny individual pixel blocks when zooming on a single grid tile using the table.

The script and functions I used to generate these (valid) grid views are in the same branch.

**Comment by martinschorb** [comment]
and it takes forever to load the grid and I am now getting these:

```
Fetched 63 image source(s) in 4686 ms, using 8 thread(s).
Transformed 235 image source(s) in 7 ms, using 8 thread(s).
Opening table:
https://raw.githubusercontent.com/mobie/centrioles-tomo-datasets/grid-test2/data/tomo/tables/crop_test1/default.tsv
Error loading setup0/timepoint0/s0 at block [5, 4, 2]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [5, 5, 2]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [6, 6, 7]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [4, 3, 1]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [6, 3, 5]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [5, 4, 0]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [4, 3, 6]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [4, 4, 1]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [5, 5, 4]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
Error loading setup0/timepoint0/s0 at block [4, 7, 2]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
```

Is there a way to find out which exact source throws these errors? Then I could check the respective n5 for the tomogram.

**Comment by tischi** [comment]
@KateMoreva Can we modify the code that throws above errors to include the name of the image source?

**Comment by KateMoreva** [comment]
> @KateMoreva Can we modify the code that throws above errors to include the name of the image source?

Yes, I'll modify it.

**Comment by tischi** [comment]
@martinschorb I am not sure, but it could be that the data is corrupt. Maybe you could also ask Josep from IT whether he has an idea. They are currently migrating the S3 storage to newer hardware, maybe that could cause hickups? You would have to tell him which bucket you are accessing.

**Comment by constantinpape** [comment]
@tischi related to the memory issue:
it looks like the volume is loaded at full scale (request for `setup0/timepoint0/s0`). This might explain bad performance and memory issue....

> Error loading setup0/timepoint0/s0 at block [4, 7, 2]: com.amazonaws.SdkClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.

**Comment by martinschorb** [comment]
When I try to load the full grid of tomograms on a VM (gave it 3 hours to fetch the data), still nothing happens and also there are no errors thrown. With some eye squinting you can see individual pixels when opening a single tomogram from the table.

**Comment by martinschorb** [comment]
The sources are shown in the BDV sources list, but nothing is shown.

**Comment by martinschorb** [comment]
Even if I switch BDV to "single" source mode and focus on an individual tile, nothing is shown (groups don't help either).

**Comment by martinschorb** [comment]
Oh wow, when I click the "focus" button in MoBIE, it shows one of the tomograms (#30). So the data is in BDV!
All other tomograms though are invisble or completely misplaced...

**Comment by martinschorb** [comment]
Also, the color overlay of the grid is not shown.

**Comment by martinschorb** [comment]
Since also the coloured grid is only shown in that one position, could it be that the location of the tiles inside the grid is off?

**Comment by martinschorb** [comment]
The one source that is shown successfully is the very first one in the BDV list.

**Comment by martinschorb** [comment]
The tiles are in fact totally displaced!

**Comment by martinschorb** [comment]
https://user-images.githubusercontent.com/35071867/136025484-06066b07-25b2-4537-adcb-81462d35fc37.mp4

**Comment by tischi** [comment]
We had this working for two cropped source, didn't we?

Could you make view with with, e.g., only three cropped sources?
That would be easier to debug, especially when working from home with slow internet.

**Comment by martinschorb** [comment]
it works for cropped sources (slowly for many)  - see `centriole-grid` and `crop_test1`views.

These are now many single tomograms. The view is generated with https://github.com/mobie/centrioles-tomo-datasets/blob/270d18903fb498d3beef491d0e0e8c720101c726/add_grid_view.py#L5 .

@constantinpape: it seems that somewhere in the grid view generation the positions are screwed up.

**Comment by martinschorb** [comment]
Now, when I try with just a couple of tomograms with that function, I run into the cell size error again.

I will try again using the `transformGrid` view type.

**Comment by martinschorb** [comment]
still, the grid tiles are miles apart...

**Comment by constantinpape** [comment]
> @constantinpape: it seems that somewhere in the grid view generation the positions are screwed up.

I am not sure what you mean; the grid transform in the spec does not store any positions; this is all done by the viewer internally.

One thing to note, also related to #479, is that for n >= 10 the sources in `sourceAnnotationDisplay` in the spec will be string sorted (i.e. 0, 1, 10, 100, ...).
This would be hard to change (because keys are always string in json) unless we switch `sourceAnnotationDisplay` to a list. In theory the order in there should not matter though, because this should be loaded as (unordered) map.

Otherwise I am not sure how we can debug this best with the current information, it would be good to have a branch with the EXACT function that generates the view and the view that fails and nothing else.

**Comment by martinschorb** [comment]
OK, a clean branch does not have these issues...

check  out `grid-test-tomo`...

Unfortunately my laptop cannot load all 1460 sources in parallel:

```
Exception in thread "pool-14-thread-5" java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:717)
	at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:957)
	at java.util.concurrent.ThreadPoolExecutor.processWorkerExit(ThreadPoolExecutor.java:1025)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

I will try on a VM.

**Comment by tischi** [comment]
@constantinpape 

I tried opening the view `grid: B-ALL-tomos` from the branch `grid-test2` from https://github.com/mobie/centrioles-tomo-datasets

And for me this works (within a few seconds even from home):

<img width="484" alt="image" src="https://user-images.githubusercontent.com/2157566/136048505-7b0d4cf4-97a4-4914-a655-7345a9aad933.png">
 
Also the annotation id sorting is numerical:

<img width="624" alt="image" src="https://user-images.githubusercontent.com/2157566/136048630-2dcdddda-fb05-463b-99ca-c67be0078329.png">

@martinschorb what exactly is not working for you?

**Comment by martinschorb** [comment]
> @martinschorb what exactly is not working for you?

@constantinpape 's suggestion of a clean branch without all the test grid views seems to have done it...

Now, we have to find out how to overcome the memory issue to get a nice figure with all tomograms in a single grid 🔢

**Comment by tischi** [comment]
> Now, we have to find out how to overcome the memory issue to get a nice figure with all tomograms in a single grid

Which memory issue?

**Comment by martinschorb** [comment]
when loading all 1460 tomos in one grid (view: `all-tomos`)

**Comment by tischi** [comment]
For me there is no such view:

<img width="541" alt="image" src="https://user-images.githubusercontent.com/2157566/136050259-f551acf9-707f-441c-a1ce-01beb679fb59.png">

**Comment by martinschorb** [comment]
branch: `grid-test-tomo`

**Comment by martinschorb** [comment]
for me the initial table is sorted alphabetically. Once I request it to be sorted by ID, it's numerical.

![Screenshot 2021-10-05 at 17 09 04](https://user-images.githubusercontent.com/35071867/136050833-11f12b4b-7b03-4f01-9377-d4942caef235.png)

**Comment by tischi** [comment]
For me, this is not an issue (maybe OS dependent); anyway that's another issue.
Let's only discuss the memory thing here. Sorry, that I mixed things!

**Comment by tischi** [comment]
I am now trying to open the 1460 tomos. But I am not sure this will work.
This is exactly the use-case that we do not support yet: many sources that are individually big. Maybe we just can't do this right now.
Then you would have to split them up in smaller grids. 

I could reproduce the issue. It is not really an out-of-memory but an out-of-threads issue (which in Java confusingly is also called out-of-memory). 
I will try to find out where the threads are generated.

Anyway, as said above, this use-case my not work yet very well anyway and I would thus not focus on this.

**Comment by martinschorb** [comment]
The VM with 100GB of Java memory just took 64 seconds to load all sources.
Still no grid displayed yet...

**Comment by martinschorb** [comment]
but in terms of a fancy figure (or poster). Here's a 8K screenshot of one of the strains:

https://oc.embl.de/index.php/s/xvTZgMuUOtyrxlY

**Comment by tischi** [comment]
As I said the RAM is not the issue, it is the number of threads.

@KateMoreva we need your help 😃 

In below code (`SpimDataOpener`  in `mobie-viewer-fiji`) we also need to enable the `sharedQueue` for the case `BdvN5S3`. Could you please look into this?

```
public SpimData openSpimData(String imagePath, ImageDataFormat imageDataFormat, SharedQueue sharedQueue ) {
        SpimData spimData = null;
        switch ( imageDataFormat ) {
            case BdvN5:
            case BdvN5S3:
                spimData = BdvUtils.openSpimData( imagePath );
                break;
            case OmeZarr:
                spimData = openOmeZarrData( imagePath, sharedQueue );
                break;
```

**Comment by martinschorb** [comment]
woop woop, it's coming....

![Screenshot 2021-10-05 at 17 32 01](https://user-images.githubusercontent.com/35071867/136054828-cb731536-737f-4df3-910b-7a84ec8d19e6.png)

**Comment by martinschorb** [comment]
1468 tomograms, on average a bit less than 2GB each, so a total of 3TB tomography data inside one BDV window ;-)

**Comment by martinschorb** [comment]
![all_tomo_grid2](https://user-images.githubusercontent.com/35071867/136058882-92845686-91dd-404d-8581-24ea1ad5a704.png)

**Comment by tischi** [comment]
OMG! 😲

**Comment by tischi** [comment]
@KateMoreva @K-Meech 
I don't know how to do this and I asked on the forum: https://forum.image.sc/t/sharedqueue-for-spimdata/58484
If you happen to know the answer please also answer on the forum 😃

**Comment by constantinpape** [comment]
@tischi I talked to @KateMoreva yesterday and she will look into this.

**Comment by KateMoreva** [comment]
> @tischi I talked to @KateMoreva yesterday and she will look into this.

> @KateMoreva @K-Meech I don't know how to do this and I asked on the forum: https://forum.image.sc/t/sharedqueue-for-spimdata/58484 If you happen to know the answer please also answer on the forum smiley

Yes, I am looking into it.

**Comment by tischi** [comment]
@martinschorb can we close this issue?

**Comment by martinschorb** [comment]
I guess it's OK for now. Obviously, improvements in performance are always nice to have and you will notice it the most with these views that contain many sources and transformations.
